src/Mutation.ts

```diff
@@ -0,0 +1,86 @@
+import * as React from 'react';
+import * as PropTypes from 'prop-types';
+import ApolloClient from 'apollo-client';
+const invariant = require('invariant');
+import { DocumentNode } from 'graphql';
+
+import { OperationVariables } from './types';
+
+export interface MutationProps {
+  mutation: DocumentNode;
+  variables?: OperationVariables;
+  children: (mutateFn: () => void, result?: any) => React.ReactNode;
+}
+
+export interface MutationState {
+  notCalled: boolean;
+  error?: string;
+  data?: any;
+  loading?: boolean;
+}
+
+class Mutation extends React.Component<MutationProps, MutationState> {
+  private client: ApolloClient<any>;
+
+  static contextTypes = {
+    client: PropTypes.object.isRequired,
+  };
+
+  constructor(props: any, context: any) {
+    super(props, context);
+
+    invariant(
+      !!context.client,
+      `Could not find "client" in the context of Mutation. Wrap the root component in an <ApolloProvider>`,
+    );
+    this.client = context.client;
+
+    this.state = {
+      notCalled: true,
+    };
+  }
+
+  private runMutation = async () => {
+    const { mutation, variables } = this.props;
+    this.setState({
+      loading: true,
+      error: undefined,
+      data: undefined,
+      notCalled: false,
+    });
+
+    try {
+      const data = this.client.mutate({
+        mutation,
+        variables,
+      });
+
+      this.setState({
+        loading: false,
+        data,
+      });
+    } catch (e) {
+      this.setState({
+        loading: false,
+        error: e,
+      });
+    }
+  };
+
+  render() {
+    const { children } = this.props;
+    const { loading, data, error, notCalled } = this.state;
+
+    const result = notCalled
+      ? undefined
+      : {
+          loading,
+          data,
+          error,
+        };
+
+    return children(this.runMutation, result);
```

Comment:
What if we add props instead of result here? That way people can still use polling functions from the component as well.